File: Core/src/main/java/com/inverce/mod/core/Ui.java
package com.inverce.mod.core;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.graphics.drawable.StateListDrawable;
import android.os.Build;
import android.os.Looper;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.util.StateSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.view.ViewTreeObserver;
import android.view.inputmethod.InputMethodManager;

import java.text.Normalizer;

/**
 * The Ui utilities
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class Ui {
    @NonNull
    private static String PADDING_NPE = "To set padding you must provide VIEW";

    /**
     * Is on ui thread.
     *
     * @return the boolean
     */
    public static boolean isUiThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }

    /**
     * Make selector state for specified colors
     *
     * @param drawable    the drawable
     * @param pressedRes  the pressed res
     * @param disabledRes the disabled res
     * @return the state list drawable
     */
    @NonNull
    public static StateListDrawable makeSelector(@NonNull Drawable drawable, @ColorRes int pressedRes, @ColorRes int disabledRes) {
        StateListDrawable state = new StateListDrawable();
        LayerDrawable pressed = new LayerDrawable(new Drawable[]{
                drawable,
                new ColorDrawable(ActivityCompat.getColor(IM.context(), pressedRes))
        });
        LayerDrawable disabled = new LayerDrawable(new Drawable[]{
                drawable,
                new ColorDrawable(ActivityCompat.getColor(IM.context(), disabledRes))
        });
        state.addState(new int[]{android.R.attr.state_pressed}, pressed);
        state.addState(new int[]{android.R.attr.state_focused}, disabled);
        state.addState(StateSet.WILD_CARD, drawable);
        return state;
    }

    /**
     * Changes visibility of view in safe and simple manner.
     *
     * @param view    the view
     * @param visible the visibility
     * @return whatever view will be visible
     */
    public static boolean visible(@Nullable View view, boolean visible) {
        return visible(view, visible, true);
    }

    /**
     * Changes visibility of view in safe and simple manner.
     *
     * @param view    the view
     * @param visible the visibility
     * @param gone    whatever use GONE or INVISIBLE when visible is false
     * @return whatever view will be visible
     */
    public static boolean visible(@Nullable View view, boolean visible, boolean gone) {
        if (view != null) {
            if (visible) {
                view.setVisibility(View.VISIBLE);
                return true;
            } else {
                if (gone) {
                    view.setVisibility(View.GONE);
                } else {
                    view.setVisibility(View.INVISIBLE);
                }
                return false;
            }
        }
        return false;
    }

    @NonNull
    public static Point getRelativePosition(@NonNull View view, @Nullable View parent) {
        Point position = new Point(view.getLeft(), view.getTop());
        ViewParent actView = view.getParent();

        if (actView == null) {
            return position;
        }
        do {
            if (actView instanceof View) {
                position.x += ((View) actView).getLeft();
                position.y += ((View) actView).getTop();
            }
            actView = actView.getParent();
        }
        while (actView instanceof View && actView != parent);
        return position;
    }

    @NonNull
    public static Point getPositionOnScreen(@NonNull View view) {
        int[] pos = new int[2];
        view.getLocationOnScreen(pos);
        return new Point(pos[0], pos[1]);
    }

    @Nullable
    public static Bitmap createScreenShoot(@Nullable Fragment view) {
        return view != null ? createScreenShoot(view.getView()) : null;
    }

    @Nullable
    public static Bitmap createScreenShoot(@Nullable View view) {
        if (view == null) {
            return null;
        }
        try {
            // create bitmap screen capture
            view.setDrawingCacheEnabled(true);
            Bitmap bitmap = Bitmap.createBitmap(view.getDrawingCache());
            view.setDrawingCacheEnabled(false);
            return bitmap;
        } catch (Throwable e) {
            // Several error may come out with file handling or OOM
            Log.exs(e);
        }

        return null;
    }

    /**
     * Run on next layout.
     *
     * @param rootView the root view
     * @param run      the run
     */
    public static void runOnNextLayout(@NonNull final View rootView, @NonNull final Runnable run) {
        rootView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    rootView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                } else {
                    //noinspection deprecation
                    rootView.getViewTreeObserver().removeGlobalOnLayoutListener(this);
                }
                run.run();
            }
        });
    }

    /**
     * Hide soft input.
     *
     * @param view the view
     */
    public static void hideSoftInput(@NonNull View view) {
        try {
            InputMethodManager imm = (InputMethodManager) IM.context().getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        } catch (Exception ignored) { /* safely ignore, as ex in here means we could not hide keyboard */ }
    }

    public static void showSoftInput(@NonNull View view) {
        showSoftInput(view, false);
    }

    public static void showSoftInput(@NonNull View view, boolean useImplicit) {
        try {
            InputMethodManager imm = (InputMethodManager) IM.context().getSystemService(Context.INPUT_METHOD_SERVICE);
            if (useImplicit) {
                imm.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT);
            } else {
                imm.toggleSoftInput(0, 0);
            }
        } catch (Exception ignored) { /* safely ignore, as ex in here means we could not hide keyboard */ }
    }

    public static String deAccent(@NonNull String name) {
        return Normalizer.normalize(name, Normalizer.Form.NFD)
                .replaceAll("ł", "l")
                .replaceAll("Ł", "L")
                .replaceAll("[^\\p{ASCII}]", "");
    }

    @SuppressWarnings("ResourceType")
    public static class Layout {
        @Nullable
        View view;
        ViewGroup.LayoutParams params;

        public Layout(@NonNull View view) {
            this.view = view;
            this.params = view.getLayoutParams();
        }

        public Layout(@NonNull ViewGroup.LayoutParams params) {
            this.params = params;
        }

        @NonNull
        public static Layout on(@NonNull View view) {
            return new Layout(view);
        }

        @NonNull
        public static Layout on(@NonNull ViewGroup.LayoutParams params) {
            return new Layout(params);
        }

        private void checkView(String exMsg) throws NullPointerException {
            if (view == null) {
                throw new NullPointerException(exMsg);
            }
        }

        @NonNull
        public Layout width(int newValue, boolean usePixels) {
            if (params != null)
                params.width = usePixels ? newValue : Screen.dpToPx(newValue);
            return this;
        }

        @NonNull
        public Layout height(int newValue, boolean usePixels) {
            if (params != null)
                params.height = usePixels ? newValue : Screen.dpToPx(newValue);
            return this;
        }


        @NonNull
        public Layout topPadding(int padding, boolean usePixels) {
            checkView(PADDING_NPE);
            assert view != null;
            view.setPadding(view.getPaddingLeft(), usePixels ? padding : Screen.dpToPx(padding), view.getPaddingRight(), view.getPaddingBottom());
            return this;
        }

        @NonNull
        public Layout leftPadding(int padding, boolean usePixels) {
            checkView(PADDING_NPE);
            assert view != null;
            view.setPadding(usePixels ? padding : Screen.dpToPx(padding), view.getPaddingTop(), view.getPaddingRight(), view.getPaddingBottom());
            return this;
        }

        @NonNull
        public Layout rightPadding(int padding, boolean usePixels) {
            checkView(PADDING_NPE);
            assert view != null;
            view.setPadding(view.getPaddingLeft(), view.getPaddingTop(), usePixels ? padding : Screen.dpToPx(padding), view.getPaddingBottom());
            return this;
        }

        @NonNull
        public Layout bottomPadding(int padding, boolean usePixels) {
            checkView(PADDING_NPE);
            assert view != null;
            view.setPadding(view.getPaddingLeft(), view.getPaddingTop(), view.getPaddingRight(), usePixels ? padding : Screen.dpToPx(padding));
            return this;
        }

        public void done() {
            if (view != null) {
                view.setLayoutParams(params);
            }
        }
    }

    @SuppressWarnings("ResourceType")
    public static class Margin extends Layout {
        ViewGroup.MarginLayoutParams params;

        Margin(@NonNull View view) {
            super(view);
            this.params = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        }

        Margin(@NonNull ViewGroup.MarginLayoutParams params) {
            super(params);
            this.params = params;
        }

        @NonNull
        public static Margin on(@NonNull View view) {
            return new Margin(view);
        }

        @NonNull
        public static Margin on(@NonNull ViewGroup.MarginLayoutParams params) {
            return new Margin(params);
        }

        @NonNull
        public Margin top(int margin, boolean usePixels) {
            params.topMargin = usePixels ? margin : Screen.dpToPx(margin);
            return this;
        }

        @NonNull
        public Margin left(int margin, boolean usePixels) {
            params.leftMargin = usePixels ? margin : Screen.dpToPx(margin);
            return this;
        }

        @NonNull
        public Margin right(int margin, boolean usePixels) {
            params.rightMargin = usePixels ? margin : Screen.dpToPx(margin);
            return this;
        }

        @NonNull
        public Margin bottom(int margin, boolean usePixels) {
            params.bottomMargin = usePixels ? margin : Screen.dpToPx(margin);
            return this;
        }

        public void done() {
            if (view != null) {
                view.setLayoutParams(params);
            }
        }
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/utils/Ui.kt
@file:Suppress("unused")

package com.inverce.mod.v2.core.utils


import android.content.Context
import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.StateListDrawable
import android.os.Build
import android.os.Looper
import android.support.annotation.ColorRes
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.util.StateSet
import android.view.View
import android.view.ViewGroup
import android.view.ViewParent
import android.view.ViewTreeObserver
import android.view.inputmethod.InputMethodManager
import com.inverce.mod.core.IM
import com.inverce.mod.v2.core.Log
import java.io.Closeable
import java.text.Normalizer

object Ui {
    private val PADDING_NPE = "To set padding you must provide VIEW"

    /**
     * Is on ui thread.
     *
     * @return the boolean
     */
    @JvmStatic val isUiThread
        get() = Looper.myLooper() == Looper.getMainLooper()

    /**
     * Make selector state for specified colors
     *
     * @param drawable    the drawable
     * @param pressedRes  the pressed res
     * @param disabledRes the disabled res
     * @return the state list drawable
     */
    @JvmStatic fun makeSelector(drawable: Drawable, @ColorRes pressedRes: Int, @ColorRes disabledRes: Int): StateListDrawable {
        val state = StateListDrawable()
        val pressed = LayerDrawable(arrayOf(drawable, ColorDrawable(ActivityCompat.getColor(IM.context(), pressedRes))))
        val disabled = LayerDrawable(arrayOf(drawable, ColorDrawable(ActivityCompat.getColor(IM.context(), disabledRes))))
        state.addState(intArrayOf(android.R.attr.state_pressed), pressed)
        state.addState(intArrayOf(android.R.attr.state_focused), disabled)
        state.addState(StateSet.WILD_CARD, drawable)
        return state
    }

    /**
     * Changes visibility of view in safe and simple manner.
     *
     * @param view    the view
     * @param visible the visibility
     * @param gone    whatever use GONE or INVISIBLE when visible is false
     * @return whatever view will be visible
     */
    @JvmStatic @JvmOverloads
    fun View?.visible(visible: Boolean, gone: Boolean = true): Boolean {
        when {
            this == null -> return false
            visible -> visibility = View.VISIBLE
            gone -> visibility = View.GONE
            else -> visibility = View.INVISIBLE
        }
        return visible
    }

    @JvmStatic
    fun View.getRelativePosition(parent: View?): Point {
        val position = Point(this.left, this.top)
        var actView: ViewParent? = this.parent ?: return position
        do {
            if (actView is View) {
                position.x += (actView as View).left
                position.y += (actView as View).top
            }
            actView = actView!!.parent
        } while (actView is View && actView !== parent)
        return position
    }

    @JvmStatic
    fun View.getPositionOnScreen(): Point {
        val pos = IntArray(2)
        this.getLocationOnScreen(pos)
        return Point(pos[0], pos[1])
    }

    @JvmStatic
    fun Fragment.createScreenShoot(): Bitmap? {
        return this.view?.createScreenShoot()
    }

    @JvmStatic
    fun View?.createScreenShoot(): Bitmap? = this?.let {
        try {
            isDrawingCacheEnabled = true
            val bitmap = Bitmap.createBitmap(it.drawingCache)
            isDrawingCacheEnabled = false
            bitmap
        } catch (e: Throwable) {
            Log.exs(e)
            null
        }
    }

    /**
     * Run on next layout.
     *
     * @param rootView the root view
     * @param run      the run
     */
    @JvmStatic
    fun runOnNextLayout(rootView: View, run: Runnable) {
        rootView.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    rootView.viewTreeObserver.removeOnGlobalLayoutListener(this)
                } else {
                    @Suppress("DEPRECATION")
                    rootView.viewTreeObserver.removeGlobalOnLayoutListener(this)
                }
                run.run()
            }
        })
    }

    /**
     * Hide soft input.
     *
     * @param view the view
     */
    @JvmStatic
    fun hideSoftInput(view: View) {
        try {
            val imm = IM.context().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        } catch (ignored: Exception) {
            /* safely ignore, as ex in here means we could not hide keyboard */
        }
    }

    @JvmStatic @JvmOverloads
    fun showSoftInput(view: View, useImplicit: Boolean = false) {
        try {
            val imm = IM.context().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            when {
                useImplicit -> imm.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT)
                else -> imm.toggleSoftInput(0, 0)
            }
        } catch (ignored: Exception) {
            /* safely ignore, as ex in here means we could not hide keyboard */
        }
    }

    @JvmStatic
    fun deAccent(name: String): String {
        return Normalizer.normalize(name, Normalizer.Form.NFD)
                .replace("ł".toRegex(), "l")
                .replace("Ł".toRegex(), "L")
                .replace("[^\\p{ASCII}]".toRegex(), "")
    }

    @Suppress("UNCHECKED_CAST")
    open class Layout<P: ViewGroup.LayoutParams>(val params: P, val view: View? = null) : Closeable {
        constructor(view: View) : this(view.layoutParams as P, view)

        @JvmOverloads
        fun setLayoutSize(width: Int = 0, height: Int = 0) {
            params.width = width
            params.height = height
        }

        @JvmOverloads
        fun setPadding(left: Int = 0, top: Int = 0, right: Int = 0, bottom: Int = 0) {
            view?.setPadding(left, top, right, bottom)
            view?.setPadding() = 33
        }
        override fun close() {
            view?.layoutParams = params
        }

        companion object {
            fun on(view: View): Layout<ViewGroup.LayoutParams> = Layout(view)
            fun on(params: ViewGroup.LayoutParams) = Layout(params)
        }
    }


//    class Margin : Layout2 {
//        internal var params: ViewGroup.MarginLayoutParams
//
//        internal constructor(view: View) : super(view) {
//            this.params = view.layoutParams as ViewGroup.MarginLayoutParams
//        }
//
//        internal constructor(params: ViewGroup.MarginLayoutParams) : super(params) {
//            this.params = params
//        }
//
//        fun top(margin: Int, usePixels: Boolean): Margin {
//            params.topMargin = if (usePixels) margin else Screen.dpToPx(margin.toFloat())
//            return this
//        }
//
//        fun left(margin: Int, usePixels: Boolean): Margin {
//            params.leftMargin = if (usePixels) margin else Screen.dpToPx(margin.toFloat())
//            return this
//        }
//
//        fun right(margin: Int, usePixels: Boolean): Margin {
//            params.rightMargin = if (usePixels) margin else Screen.dpToPx(margin.toFloat())
//            return this
//        }
//
//        fun bottom(margin: Int, usePixels: Boolean): Margin {
//            params.bottomMargin = if (usePixels) margin else Screen.dpToPx(margin.toFloat())
//            return this
//        }
//
//        override fun done() {
//            if (view != null) {
//                view!!.layoutParams = params
//            }
//        }
//
//        companion object {
//
//            fun on(view: View): Margin {
//                return Margin(view)
//            }
//
//            fun on(params: ViewGroup.MarginLayoutParams): Margin {
//                return Margin(params)
//            }
//        }
//    }
}